### ISSUE TITLE
Backward propagating eigenmode adjoints are wrong

### ISSUE THREAD

**Issue by smartalecH** [NanoComp/meep#1585]
Note that objective functions that depend on the backward propagating mode produce incorrect gradients right now.

We can get around this by calculating the *forward* coefficient in the *backward* direction (i.e. flip the coordinate frame of MPB) using something like this:

```python3
bw = mpa.EigenmodeCoefficient(sim, source_vol, kpoint_func=lambda a,b: mp.Vector3(x=-1))
```

Since this hack works, it's probably just some trivial sign error in the adjoint pipeline (which is also why I haven't pursued a fix yet...). That being said, we should add a unit test that verifies this (and someday actually fix this).

**Comment by stevengj** [comment]
Test case:
![image](https://user-images.githubusercontent.com/2913679/120413075-a5ee5b80-c325-11eb-9085-715aca62c2f6.png)

**Comment by ahoenselaar** [comment]
There is certainly an extremely dangerous coding practice in use here:
https://github.com/NanoComp/meep/blob/master/python/adjoint/objective.py#L34

Going from a boolean 'forward' to an integer 'self.forward' with the opposite boolean interpretation is asking for trouble.
`forward=True` becomes `self.forward=0` (interpreted as `False` in a Boolean context).

**Comment by smartalecH** [comment]
> There is certainly an extremely dangerous coding practice in use here

Agreed but not the cause of the above issue, luckily (in case anyone was wondering).

**Comment by ahoenselaar** [comment]
It would be great to have a small repro case with observed behavior and expected behavior. That could eventually be turned into a unit test.

**Comment by ianwilliamson** [comment]
> It would be great to have a small repro case with observed behavior and expected behavior. That could eventually be turned into a unit test.

See the update to the JAX-Meep wrapper test in: https://github.com/NanoComp/meep/pull/1593/commits

**Comment by ahoenselaar** [comment]
Great, thank you!
